File: src/main/services/cloud-save/path-resolution/tokens.ts
import type {
  CloudSavePathToken,
  CloudSavePathResolutionContext,
  CloudSaveTokenMap,
} from "./types";

const normalizeTokenValues = (
  values: Array<string | null | undefined>
): string[] => {
  return Array.from(
    new Set(
      values.filter((value): value is string => {
        return typeof value === "string" && value.length > 0;
      })
    )
  );
};

const addToken = (
  tokenMap: CloudSaveTokenMap,
  token: CloudSavePathToken,
  values: Array<string | null | undefined>
): void => {
  const normalizedValues = normalizeTokenValues(values);

  if (normalizedValues.length > 0) {
    tokenMap[token] = normalizedValues;
  }
};

export const buildCloudSaveTokenMap = (
  context: CloudSavePathResolutionContext
): CloudSaveTokenMap => {
  const tokenMap: CloudSaveTokenMap = {};

  addToken(tokenMap, "<base>", [context.installDir]);
  addToken(tokenMap, "<home>", [context.homeDir]);
  addToken(tokenMap, "<storeUserId>", context.steamUserIds);
  addToken(tokenMap, "<winAppData>", [context.appDataDir]);
  addToken(tokenMap, "%APPDATA%", [context.appDataDir]);
  addToken(tokenMap, "<winLocalAppData>", [context.localAppDataDir]);
  addToken(tokenMap, "%LOCALAPPDATA%", [context.localAppDataDir]);
  addToken(tokenMap, "<winDocuments>", [context.documentsDir]);
  addToken(tokenMap, "<winPublic>", [context.publicDir]);
  addToken(tokenMap, "<winProgramData>", [context.programDataDir]);

  return tokenMap;
};

File: native/hydra-native/src/cloud_save/path_resolution/tokens.rs
use super::context::normalize_separators;

const MISSING_TOKEN: &str = "<skip>";

pub struct TokenValues {
    pub root: String,
    pub base: String,
    pub home: String,
    pub os_username: String,
    pub app_data: String,
    pub local_app_data: String,
    pub documents: String,
    pub public: String,
    pub program_data: String,
    pub windows_dir: String,
    pub xdg_data: String,
    pub xdg_config: String,
}

pub fn literal(value: &str) -> String {
    globset::escape(&normalize_separators(value))
}

pub fn optional_literal(value: Option<&str>) -> String {
    value.map(|value| literal(value)).unwrap_or_else(missing)
}

pub fn missing() -> String {
    MISSING_TOKEN.to_string()
}

pub fn apply(path: &str, values: &TokenValues) -> String {
    path.replace("<root>", &values.root)
        .replace("<base>", &values.base)
        .replace("<home>", &values.home)
        .replace("<osUserName>", &values.os_username)
        .replace("<winAppData>", &values.app_data)
        .replace("%APPDATA%", &values.app_data)
        .replace("<winLocalAppData>", &values.local_app_data)
        .replace("%LOCALAPPDATA%", &values.local_app_data)
        .replace("<winDocuments>", &values.documents)
        .replace("<winPublic>", &values.public)
        .replace("<winProgramData>", &values.program_data)
        .replace("<winDir>", &values.windows_dir)
        .replace("<xdgData>", &values.xdg_data)
        .replace("<xdgConfig>", &values.xdg_config)
}

pub fn expand_store_user_ids(path: String, steam_user_ids: &[String]) -> Vec<String> {
    if !path.contains("<storeUserId>") {
        return vec![path];
    }

    steam_user_ids
        .iter()
        .map(|user_id| path.replace("<storeUserId>", &globset::escape(user_id)))
        .collect()
}

pub fn has_unresolved_placeholder(path: &str) -> bool {
    path.contains('<')
}

pub fn tokens_in_path(path: &str) -> Vec<String> {
    let mut tokens = Vec::new();
    let mut rest = path;

    while let Some(start) = rest.find('<') {
        let after = &rest[start..];
        let Some(end) = after.find('>') else {
            break;
        };

        let token = &after[..=end];
        if !tokens.iter().any(|existing| existing == token) {
            tokens.push(token.to_string());
        }
        
        rest = &after[end + 1..];
    }

    for token in ["%APPDATA%", "%LOCALAPPDATA%"] {
        if path.contains(token) && !tokens.iter().any(|existing| existing == token) {
            tokens.push(token.to_string());
        }
    }

    tokens
}
